logger/handlers/ringbuffer.go

```diff
@@ -0,0 +1,134 @@
+package handlers
+
+import (
+	"container/ring"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/syslog"
+)
+
+var projectBuffer = make(map[string]*ring.Ring)
+var ringBufferSize int
+
+func addToBuffer(name string, message string) {
+	currentRing, ok := projectBuffer[name]
```

Comment:
I think there may be a race condition here, too. It is possible for projectBuffer to be read/written-to out of order. (viz. A reads, ok = false. B reads, ok= false. A creates a new ring ring and adds to buffer. B creates new ring and adds to buffer. A's ring is lost.